clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -3439,10 +3486,15 @@ void Sema::MarkDevice(void) {
     }
   }
 
+  llvm::SmallPtrSet<FunctionDecl *, 10> SYCLFunctionCyclicCheckSet;
   for (Decl *D : syclDeviceDecls()) {
     if (auto SYCLKernel = dyn_cast<FunctionDecl>(D)) {
       llvm::SmallPtrSet<FunctionDecl *, 10> VisitedSet;
-      Marker.CollectKernelSet(SYCLKernel, SYCLKernel, VisitedSet);
+      llvm::SmallPtrSet<FunctionDecl *, 10> SYCLKernelInvokeSet;
+      Marker.WalkSYCLFunctionCG(SYCLKernel, SYCLKernelInvokeSet,
+                                SYCLFunctionCyclicCheckSet);
+      Marker.SYCLKernelInvokeMap.insert(
```

Comment:
It seems to me the ownership model is all wrong here.  The Marker type I think should just own these operations and collections.  We shouldn't have to initialize 4 collections for it.